src/v/wasm/wasmtime_async.cc

```diff
@@ -287,19 +299,57 @@ class wasmtime_engine : public engine {
     }
 
 private:
-    void reset_fuel() {
-        auto* ctx = wasmtime_store_context(_store.get());
+    void reset_fuel(wasmtime_context_t* ctx) {
         uint64_t fuel = 0;
         wasmtime_context_fuel_remaining(ctx, &fuel);
         handle<wasmtime_error_t, wasmtime_error_delete> error(
           wasmtime_context_add_fuel(ctx, wasmtime_fuel_amount - fuel));
     }
 
+    ss::future<> create_instance() {
+        // The underlying "data" for this store is our engine, which is what
+        // allows our host functions to access the actual module for that host
+        // function.
+        handle<wasmtime_store_t, wasmtime_store_delete> store{
+          wasmtime_store_new(_engine, /*data=*/this, /*finalizer=*/nullptr)};
+        auto* context = wasmtime_store_context(store.get());
+        reset_fuel(context);
+
+        _wasi_module.set_timestamp(model::timestamp::min());
+
+        // The wasm spec has a feature that a module can specify a startup
+        // function that is run on start.
+        wasm_trap_t* trap_ptr = nullptr;
+        wasmtime_error_t* error_ptr = nullptr;
+        handle<wasmtime_call_future_t, wasmtime_call_future_delete> fut{
+          wasmtime_instance_pre_instantiate_async(
+            _preinitialized->underlying.get(),
+            context,
+            &_instance,
+            &trap_ptr,
+            &error_ptr)};
+
+        // Poll the call future to completion, yielding to the scheduler when
+        // the future yields.
+        while (!wasmtime_call_future_poll(fut.get())) {
+            co_await ss::coroutine::maybe_yield();
```

Comment:
From https://docs.wasmtime.dev/api/wasmtime/struct.Config.html#asynchronous-wasm

> The poll method of the futures returned by Wasmtime will perform the actual work of calling the WebAssembly. Wasmtime won’t manage its own thread pools or similar, that’s left up to the embedder.
> To implement futures in a way that WebAssembly sees asynchronous host functions as synchronous, all async Wasmtime futures will execute on a separately allocated native stack from the thread otherwise executing Wasmtime. This separate native stack can then be switched to and from. Using this whenever an async host function returns a future that resolves to Pending we switch away from the temporary stack back to the main stack and propagate the Pending status.

Basically calling poll here switches to the "Wasm" stack and runs the VM. Host functions that are async end up being propagated to `poll` and the `continuation` we return in async host functions is the bit that checks if a `poll` is able to go back to executing Wasm within the VM. This is why we need to pass back the future from continutation otherwise we'll busy loop asking if the future is done, instead of letting seastar reschedule this fiber when the future is done (at that point the continuation will return true and go back to running the VM).

Reply:
awesome